pricegraph/src/orderbook.rs

```diff
@@ -271,6 +270,56 @@ impl Orderbook {
         Some(invert_price(last_transient_price))
     }
 
+    /// Fill an order at a given price, returning the estimated buy amount that
+    /// can be filled at the given price with the current existing orders.
+    ///
+    /// This is calculated by repeatedly finding the cheapest path between a
+    /// token pair that is below the specified price and adding the capacity of
+    /// the path to the result.
+    ///
+    /// Note that the limit price is expressed as an exchange limit price, i.e.
+    /// with implicitely included fees.
+    pub fn fill_order_at_price(&mut self, pair: TokenPair, actual_limit_price: f64) -> f64 {
+        if !self.is_token_pair_valid(pair) {
+            return 0.0;
+        }
+        self.update_projection_graph();
+
+        let (sell, buy) = (node_index(pair.sell), node_index(pair.buy));
+        let effective_limit_price = actual_limit_price / FEE_FACTOR;
+
+        let mut total_volume = 0.0;
+        let mut predecessors = self.reduced_shortest_paths(sell);
+        while let Some(path) = path::find_path(&predecessors, sell, buy) {
+            let (capacity, transient_price) =
+                self.find_path_capacity_and_price(&path).unwrap_or_else(|| {
+                    panic!(
+                        "failed to fill detected shortest path {}",
+                        format_path(&path),
+                    )
+                });
+
+            if transient_price > effective_limit_price {
+                break;
+            }
```

Comment:
This relies on the order of the results in `find_path` and `find_path_capacity_and_price`, so that we get better paths first, right? Is that order guaranteed?

Reply:
Yes, we use Bellman-Ford to find the "best" path first. The search criteria is to find the path with the lowest `transient_price` from the sell token to the buy token.